internal/app/upstream/client.go

```diff
@@ -0,0 +1,57 @@
+package client
+
+import (
+	"context"
+
+	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"google.golang.org/grpc"
+)
+
+// XdsClient handles the requests and responses from the origin server.
+// The xds client handles each xds request on a separate stream,
+// e.g. 2 different cds requests happen on 2 separate streams.
+// It is the caller's responsibility to make sure there is one instance of XdsClient per unique xds request.
+type XdsClient interface {
+	// QueueRequest creates a stream with the origin server
+	// All discovery requests to origin server arrive on the request channel
+	// All responses from the origin server are sent back through the response channel
+	// The caller is responsible for managing the lifetime of the channels.
+	// QueueRequest uses the retry and timeout configurations to make best effort to get the responses from origin server.
+	// If there's a new request in between retries, the retries are abandoned.
+	// The request and response happen asynchronously. Retries are scoped for sending messages to origin server.
+	// If the timeouts are exhausted, receive fails or a irrecoverable error occurs, the error is sent back to the caller.
+	// It is the caller's responsibility to send a new request from the last known DiscoveryRequest.
+	// Cancellation and cleanup operations will be based on cancellation of the context and closing of channels.
+	QueueRequest(context.Context, chan *v2.DiscoveryRequest, chan *Response) error
```

Comment:
Thanks this makes more sense now. I think a better name for this method might be `CreateRequestResponseChannel` or similar? Since this doesn't actually queue anything?

Reply:
I like `xdsClient.Start()` or `xdsClient.Initialize()` better. Does one of these sound reasonable?

Reply:
If this is a one shot function, why does it need to live on the interface at all? Can't it be part of the constructor for the thing below? Otherwise yes either Start or Initialize seems fine.

Reply:
The scenario we might need the function is when we initialize the constructor with args at a different place than where the function called. Since we don't have the impl structure yet, it's hard to imagine, but gives a reasonable way for other parts to Mock this interface and come up with other interfaces.
Although creating a higher order `func` will be ok for this purpose, i did not want to use it yet.

Reply:
renamed the function to `Start` for now.

Reply:
> The scenario we might need the function is when we initialize the constructor with args at a different place than where the function called.

I couldn't quite follow this, can you give me an example?

Reply:
I think that there should be a top level `NewClient` function that does what's in the interface currently, and the client interface would containing a contract for adding new streams. I'm on the same page as Matt that I don't think the one shot func should live in the interface.

```
package upstream

...

type Client interface {
    AddStream(*v2.DiscoveryRequest) error
}

NewClient(context.Context, chan *Response) (Client, error)
```

The implementation of AddStream would push it into the an internal request channel maintained by the client to handle retries and what not. The caller would still maintain the response channel to receive responses in a non-blocking manner.

What are your thoughts on this approach?

Reply:
I made some changes in the interface and it is not a one shot function anymore.
I've also created a sample impl to brainstorm with. The `NewClient` will be a singleton instance and will not be per request. I am envisioning that the NewClient will hold a single grpc client connection and will manage closing the connection on shutdown.

Everything else remains the same. @jessicayuen about your proposed changes, i think we might be constrained with how grpc works for streams. Request and response have to be called on a single stream and are asynchronous and should always be called from a single goroutine each. Given that the client will be a singleton, do you think we still need to adopt the suggestion?

ref: https://github.com/envoyproxy/xds-relay/pull/36/files